plugins/oidc/access.go

```diff
@@ -10,118 +10,207 @@ import (
 	"strings"
 
 	"github.com/Kong/go-pdk"
+	"github.com/lestrrat-go/jwx/v2/jwk"
 	"github.com/lestrrat-go/jwx/v2/jwt"
 )
 
-var headerRegex = regexp.MustCompile(`^(\w+) (\S+)$`)
+const headerRegex = `^(\w+) (\S+)$`
 
-// Access is executed every time the kong gateway receives a request. Since the
+var errEmptyEndpoint = errors.New("empty discovery endpoint")
+var errDiscoveryRequestFailure = errors.New("discovery request failed")
+var errDiscoveryUnmarshalFailure = errors.New("unable to unmarshal openid connect configuration")
+
+// discoverEndpoints retrieves the userinfo and jwks endpoints from the given
+// discovery endpoint.
+// The function sends a GET request to the discovery endpoint, which should
+// respond with the openid connect configuration.
+// This configuration is a JSON object that the function decodes into a map,
+// extracting the userinfo_endpoint and jwks_uri fields as the userinfo and jwks
+// endpoints respectively.
+// In case of error, it will also join the error with suitable contextual error
+// message and return.
+// If the discovery endpoint is empty or leads to a failed request, the function
+// returns an error.
+// If the retrieval or decoding of the openid connect configuration fails, the
+// function returns an error.
+//
+// Parameters:
+//
+//	discoveryEndpoint (string): address of the discovery endpoint.
+//
+// Returns:
+//
+//	userinfoEndpoint (string): the discovered userinfo endpoint.
+//	jwksEndpoint (string): the discovered jwks endpoint.
+//	err (error): the error, if any occurred during the process.
+func discoverEndpoints(discoveryEndpoint string) (userinfoEndpoint string, jwksEndpoint string, err error) {
+	// cleanup the discovery endpoint address and remove possible whitespaces
+	discoveryEndpoint = strings.TrimSpace(discoveryEndpoint)
+	if discoveryEndpoint == "" {
+		return "", "", errEmptyEndpoint
+	}
+
+	// now request the openid connect configuration endpoint
+	res, err := http.Get(discoveryEndpoint)
+	if err != nil {
+		return "", "", errors.Join(errDiscoveryRequestFailure, err)
+	}
+
+	// now read the openid connect configuration
+	var openIdConnectConfiguration map[string]interface{}
+	err = json.NewDecoder(res.Body).Decode(&openIdConnectConfiguration)
+	if err != nil {
+		return "", "", errors.Join(err, errDiscoveryUnmarshalFailure)
+	}
+
+	return openIdConnectConfiguration["userinfo_endpoint"].(string), openIdConnectConfiguration["jwks_uri"].(string), nil
```

Comment:
The assertion here would crash the application, is that expected behavior?
Maybe do a type check and return an error if that the json has an incorrect format instead.

Reply:
Yes, this type assertion would crash the application. A type check will be added